File: common/src/main/java/net/shiroha233/roadweaver/structures/registry/RoadsideStructureRegistry.java
package net.shiroha233.roadweaver.structures.registry;

import net.minecraft.core.Holder;
import net.minecraft.core.Registry;
import net.minecraft.core.RegistryAccess;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.levelgen.structure.Structure;
import net.shiroha233.roadweaver.structures.data.BiomeCategory;
import net.shiroha233.roadweaver.structures.types.RoadsideStructure;

import java.util.ArrayList;
import java.util.List;

/**
 * 路边结构注册中心
 */
public final class RoadsideStructureRegistry {
    private RoadsideStructureRegistry() {}
    
    private static volatile List<RoadsideStructureEntry> cache;
    
    public record RoadsideStructureEntry(
        ResourceLocation id,
        Holder<Structure> holder,
        RoadsideStructure structure
    ) {}
    
    public static List<RoadsideStructureEntry> getAll(ServerLevel level) {
        if (level == null || !Level.OVERWORLD.equals(level.dimension())) return List.of();
        List<RoadsideStructureEntry> current = cache;
        if (current != null) return current;
        synchronized (RoadsideStructureRegistry.class) {
            if (cache == null) cache = List.copyOf(loadFromRegistry(level.registryAccess()));
            return cache;
        }
    }
    
    public static RoadsideStructureEntry choose(ServerLevel level,
                                                 BiomeCategory biome,
                                                 int roadLength,
                                                 RandomSource random) {
        List<RoadsideStructureEntry> all = getAll(level);
        List<RoadsideStructureEntry> candidates = new ArrayList<>();
        int totalWeight = 0;
        
        for (RoadsideStructureEntry entry : all) {
            RoadsideStructure structure = entry.structure();
            
            if (!structure.placementRule().isBiomeAllowed(biome)) {
                continue;
            }
            
            if (!structure.placementRule().isRoadLongEnough(roadLength)) {
                continue;
            }
            
            int weight = structure.weight();
            if (weight <= 0) {
                continue;
            }
            
            candidates.add(entry);
            totalWeight += weight;
        }
        
        if (candidates.isEmpty() || totalWeight <= 0) {
            return null;
        }
        
        int roll = random.nextInt(totalWeight);
        int sum = 0;
        for (RoadsideStructureEntry entry : candidates) {
            sum += entry.structure().weight();
            if (roll < sum) {
                return entry;
            }
        }
        
        return candidates.get(0);
    }
    
    private static List<RoadsideStructureEntry> loadFromRegistry(RegistryAccess registryAccess) {
        List<RoadsideStructureEntry> result = new ArrayList<>();
        
        Registry<Structure> structureRegistry = registryAccess.registryOrThrow(Registries.STRUCTURE);
        
        for (var entry : structureRegistry.entrySet()) {
            ResourceLocation id = entry.getKey().location();
            Structure structure = entry.getValue();
            
            if (structure instanceof RoadsideStructure roadsideStructure) {
                Holder<Structure> holder = structureRegistry.getHolderOrThrow(entry.getKey());
                result.add(new RoadsideStructureEntry(id, holder, roadsideStructure));
            }
        }
        
        return result;
    }
    
    public static void clearCache() {
        cache = null;
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/structures/registry/RoadsideStructureRegistry.kt
package net.shiroha233.roadweaver.structures.registry

import net.minecraft.core.Holder
import net.minecraft.core.Registry
import net.minecraft.core.RegistryAccess
import net.minecraft.core.registries.Registries
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.level.levelgen.structure.Structure
import net.shiroha233.roadweaver.structures.data.BiomeCategory
import net.shiroha233.roadweaver.structures.types.RoadsideStructure
import java.util.concurrent.ConcurrentHashMap

// 路边结构注册中心
//
// 从世界的 Structure 注册表中读取所有 RoadsideStructure，
// 提供根据条件筛选和选择结构的方法。
//
// 数据来源：datapack 中的 worldgen/structure/*.json
object RoadsideStructureRegistry {
    // 缓存：每个世界的路边结构列表
    private val CACHE: MutableMap<ResourceKey<*>, List<RoadsideStructureEntry>> = ConcurrentHashMap()

    /**
     * 结构条目，包含 Holder 引用和解析后的结构
     */
    data class RoadsideStructureEntry(
        val id: ResourceLocation,
        val holder: Holder<Structure>,
        val structure: RoadsideStructure
    )

    /**
     * 获取所有已注册的路边结构
     */
    @JvmStatic
    fun getAll(level: ServerLevel): List<RoadsideStructureEntry> {
        val dimensionKey: ResourceKey<*> = level.dimension()
        return CACHE.computeIfAbsent(dimensionKey) { loadFromRegistry(level.registryAccess()) }
    }

    /**
     * 根据条件选择一个路边结构
     *
     * @return 选中的结构，如果没有符合条件的返回 null
     */
    @JvmStatic
    fun choose(level: ServerLevel, biome: BiomeCategory, roadLength: Int, random: RandomSource): RoadsideStructureEntry? {
        val all = getAll(level)
        val candidates = ArrayList<RoadsideStructureEntry>()
        var totalWeight = 0

        for (entry in all) {
            val structure = entry.structure

            // 群系过滤
            if (!structure.placementRule().isBiomeAllowed(biome)) {
                continue
            }

            // 道路长度过滤
            if (!structure.placementRule().isRoadLongEnough(roadLength)) {
                continue
            }

            val weight = structure.weight()
            if (weight <= 0) {
                continue
            }

            candidates.add(entry)
            totalWeight += weight
        }

        if (candidates.isEmpty() || totalWeight <= 0) {
            return null
        }

        // 权重随机选择
        val roll = random.nextInt(totalWeight)
        var sum = 0
        for (entry in candidates) {
            sum += entry.structure.weight()
            if (roll < sum) {
                return entry
            }
        }

        return candidates[0]
    }

    /**
     * 从注册表加载所有路边结构
     */
    private fun loadFromRegistry(registryAccess: RegistryAccess): List<RoadsideStructureEntry> {
        val result = ArrayList<RoadsideStructureEntry>()
        val structureRegistry: Registry<Structure> = registryAccess.registryOrThrow(Registries.STRUCTURE)

        for (entry in structureRegistry.entrySet()) {
            val id = entry.key.location()
            val structure = entry.value

            // 只收集 RoadsideStructure 类型
            val roadsideStructure = structure as? RoadsideStructure ?: continue
            val holder = structureRegistry.getHolderOrThrow(entry.key)
            result.add(RoadsideStructureEntry(id, holder, roadsideStructure))
        }

        return result
    }

    /**
     * 清除缓存（在世界卸载或重载时调用）
     */
    @JvmStatic
    fun clearCache() {
        CACHE.clear()
    }

    /**
     * 清除指定维度的缓存
     */
    @JvmStatic
    fun clearCache(dimensionKey: ResourceKey<*>) {
        CACHE.remove(dimensionKey)
    }
}
